### PR TITLE
Refactor docker_base to use inversion-of-control, and GCP to use that

### PR DESCRIPTION
**Pull request by lathanh**
## Pull Request Description

Use [inversion of control](https://en.wikipedia.org/wiki/Inversion_of_control) in `docker_base.prep_batches()` so that it can more strongly and easily ensure consistency between various implementations. Also refactor the GCP implementation to use that (the updated `docker_base.prep_batches()`).

## Checklist

Not all may apply

- [x] Code changes (must work)
- [x] Tests exercising your feature/bug fix (check coverage report on Checks -> BuildStockBatch Tests -> Artifacts)
- [x] Coverage has increased or at least not decreased. Update `minimum_coverage` in `.github/workflows/coverage.yml` as necessary.
- [x] All other unit and integration tests passing
- [x] ~~Update validation for project config yaml file changes~~
- [x] ~~Update existing documentation~~
- [x] ~~Run a small batch run on Eagle to make sure it all works if you made changes that will affect Eagle~~
- [x] ~~Add to the changelog_dev.rst file and propose migration text in the pull request~~

### CODE REVIEW

File: buildstockbatch/gcp/gcp.py [hunk 1]
```diff
@@ -429,133 +445,10 @@ def run_batch(self):
...
-
-        logger.info("Copying weather files on GCS")
-        bucket = self.gcs_bucket
-        Parallel(n_jobs=-1, verbose=5)(delayed(copy_GCS_file)(bucket, src, bucket, dest) for src, dest in epws_to_copy)
+        # Prepare batches of work
+        batch_info = self.prep_batches()
 
-        # Step 3: Define and run the GCP Batch job.
+        # Define and run the GCP Batch job.
         logger.info("Setting up GCP Batch job")
```
Comment (nweires) [inline buildstockbatch/gcp/gcp.py:437]: Do you plan to pull even more of this into docker_base.py? I was imagining `run_batch()` being in `DockerBatchBase`, where it would call `prep_batches()` and something like `self.run_cloud_batch_job()` that sets up and starts the job on GCP/AWS.

File: buildstockbatch/cloud/docker_base.py [hunk 2]
```diff
@@ -180,6 +236,56 @@ def prep_batches(self, tmppath):
         logger.debug("Done compressing job jsons using gz {:.1f} seconds".format(tick))
         shutil.rmtree(jobs_dir)
 
-        os.makedirs(tmppath / "results" / "simulation_output")
+        return DockerBatchBase.BatchInfo(n_sims=n_sims, n_sims_per_job=n_sims_per_job, job_count=job_count)
+
+    def prep_batches(self):
```
Comment (nweires) [inline buildstockbatch/cloud/docker_base.py]: There are now a lot of `prep_*` methods here that all sound similar and are getting a bit hard to follow.

Maybe `prep_batches()` could be renamed something like `upload_job_files()`, now that it's doing the uploads? And/or can we make more of the methods private (with a leading `_`) to make it a little easier to tell which of these should ultimately be called by `run_batch()`?

File: buildstockbatch/test/test_docker_base.py [hunk 3]
```diff
@@ -27,20 +27,23 @@ def test_prep_batches(basic_residential_project_file, mocker):
...
 
-    with tempfile.TemporaryDirectory(prefix="bsb_") as tmpdir, tempfile.TemporaryDirectory(
-        prefix="bsb_"
-    ) as tmp_weather_dir:
-        dbb._weather_dir = tmp_weather_dir
+    with tempfile.TemporaryDirectory(prefix="bsb_") as tmpdir:
         tmppath = pathlib.Path(tmpdir)
-
-        job_count, unique_epws = dbb.prep_batches(tmppath)
+        epws_to_copy, n_sims, job_count = dbb._prep_batch_files(tmppath)
```
Comment (nweires) [inline buildstockbatch/test/test_docker_base.py]: Also check that `n_sims` is correct.

File: buildstockbatch/test/test_docker_base.py [hunk 4]
```diff
@@ -27,20 +27,23 @@ def test_prep_batches(basic_residential_project_file, mocker):
...
+        #   * "G2500210.epw" is unique; check for it (gz'd) in tmppath
+        #   * "G2601210.epw" and "G2601390.epw" are dupes. One should be in
+        #     tmppath; one should be copied to the other according to ``epws_to_copy``
+        assert os.path.isfile(tmppath / "weather" / "G2500210.epw.gz")
+        assert os.path.isfile(tmppath / "weather" / "G2601210.epw.gz") or os.path.isfile(
+            tmppath / "weather" / "G2601390.epw.gz"
+        )
+        src, dest = epws_to_copy[0]
+        assert src == "G2601210.epw.gz" or src == "G2601390.epw.gz"
+        assert dest == "G2601210.epw.gz" or dest == "G2601390.epw.gz"
```
Comment (nweires) [inline buildstockbatch/test/test_docker_base.py]:
```suggestion
        assert src in ("G2601210.epw.gz", "G2601390.epw.gz")
        assert dest in ("G2601210.epw.gz", "G2601390.epw.gz")
```
nit: this feels more Pythonic to me

File: buildstockbatch/gcp/gcp.py (hunk 1 above)
Comment (lathanh) [reply buildstockbatch/gcp/gcp.py:437]: I was on the fence, as it was a marginal benefit to go this far, but I've gone ahead and done it.

File: buildstockbatch/cloud/docker_base.py (hunk 2 above)
Comment (lathanh) [reply buildstockbatch/cloud/docker_base.py]: This is somewhat intentional to tie them all together (they're all doing some of the prep work). I've renamed the top-level one to distinguish it a bit better, made the others private, and reordered the functions which should help make it a bit more natural to understand their relationship.

File: buildstockbatch/cloud/docker_base.py [hunk 5]
```diff
@@ -56,30 +71,164 @@ def docker_image(self):
...
-            unique_epws: A dictionary mapping from the hash of weather files to a list of filenames
-                with that hashed value. Only the first in each list is written to tmppath, so
-                this can be used to recreate the other files later.
+        :param batch_info: A :class:`DockerBatchBase.BatchInfo` containing information about the job
         """
-        # Generate buildstock.csv
-        buildstock_csv_filename = self.sampler.run_sampling()
+        raise NotImplementedError
+
+    def run_batch(self):
```
Comment (nweires) [inline buildstockbatch/cloud/docker_base.py:101]: This should probably have a docstring.

File: buildstockbatch/cloud/docker_base.py [hunk 6]
```diff
@@ -56,30 +71,164 @@ def docker_image(self):
...
+
+        self.start_batch_job(batch_info)
+
+    def _run_batch_prep(self, tmppath):
+        """Do preparation for running the Batch jobs on the cloud, including producing and uploading
+        files to the cloud that the batch jobs will use.
+
+        This includes:
+            - Weather files (:func:`_prep_weather_files_for_batch`)
+            - Assets :func:`_prep_assets_for_batch`)
```
Comment (nweires) [inline buildstockbatch/cloud/docker_base.py]:
```suggestion
            - Assets (:func:`_prep_assets_for_batch`)
```

**Review by nweires — APPROVED** [review]
